Fix issues #38855, #41075, #41717, and #43912 with firewalld state and execution modules

connordelacruz opened the pull request:
### What does this PR do?
#### firewalld state module
* Fixes issue where `firewalld.present` would not propagate errors  
* Fixes issue where running `state.apply test=True` would show that `firewalld.present` made changes despite not actually changing anything
* Adds support for modular firewall configurations by adding parameters similar to `prune_services` for all items getting pruned by default
* Changes `prune_services` default value to `False`
* Added more detailed documentation to the docstring for `firewalld.present`

#### firewalld execution module
* Fixes issue where `firewalld.add_port` and  `firewalld.add_port_fwd` always enable masquerading (affects state module as well since it calls the execution module)

### What issues does this PR fix or reference?
#38855 
#41075 
#41717 
#43912 

### Previous Behavior
#### firewalld state module
* `firewalld.present` would set the result to `True` even if there was an error
* Running `state.apply test=True` would show that `firewalld.present` made changes despite not actually changing anything
* `firewalld.present` would always prune all rich rules, sources, interfaces, port forwards, ports, and blocked ICMPs 
* `firewalld.present` would prune services by default
* using `firewalld.present` to add ports or port forwards would also enable masquerading even if `masquerade` was set to false

#### firewalld execution module
* Running `firewalld.add_port` or  `firewalld.add_port_fwd` always enabled masquerading 

### New Behavior
#### firewalld state module
* `firewalld.present` now propagates errors 
* When running `state.apply test=True`, `firewalld.present` no longer shows that it made changes when it predicts that there are changes to make
* Added new parameters to `firewalld.present` similar to `prune_services` for each of the items that were previously always pruned (rich rules, sources, interfaces, port forwards, ports, and blocked ICMPs). These are set to `False` by default.
* Changed the default value of `prune_services` to `False` to be consistent with the new parameters.
* Adding ports and port forwards no longer affects masquerading

#### firewalld execution module
* Running `firewalld.add_port` or  `firewalld.add_port_fwd` no longer affects masquerading

### Tests written?

No

connordelacruz committed:
Addressed several bugs and issues with the firewalld state

Added new parameters to resolve saltstack/salt#38855 and resolve
saltstack/salt#41075:
- Added prune_ parameters similar to prune_services for each item that gets manipulated (prune_block_icmp, prune_ports, prune_port_fwd, prune_interfaces, prune_sources, and prune_rich_rules)
- Set prune_services and all new prune_ parameters to False by default
to avoid confusion with the firewalld.present function's expected
behavior
- Modified _present() to handle each of the new prune_ parameters
similarly to how prune_services is handled

Fixed saltstack/#41717:
- Removed redundant code from present() that was overwriting ret['result']. Previously, present() called another function _present() to handle the
bulk of the actual functionality. When it returned, it reloaded the
firewalld service if changes were made, updated ret, and returned it.
However, _present() already updated ret, present() was overwriting its
changes and setting ret['result'] to true regardless of what happened in
_present()
- Fixed output showing changes when test=True. Instead of listing
predicted changes in ret['changes'], ret['comment'] displays a nested
list of items that will be changed and sets ret['changes'] = {} (based
on the way pkg.py handles its output

Additionally added TODO comment to update documentation block

connordelacruz committed:
Merge remote-tracking branch 'upstream/develop' into develop

connordelacruz committed:
Started adding documentation to firewalld.present()

connordelacruz committed:
Added all standard parameters to the docblock for present()

connordelacruz committed:
Added all standard parameters to the docblock for present()

Commit:
Merge branch 'develop' into develop

cachedout commented on salt/states/firewalld.py:
```diff
@@ -153,37 +154,85 @@ def __virtual__():
+            prune_port_fwd=False,
             services=None,
-            prune_services=True,
+            prune_services=False,
```
What effect is this going to have on people who upgrade Salt with existing `firewalld` states?

connordelacruz replied:
Sorry about the delayed response, I realized I had email notifications turned off. 

Changing the default of `prune_services` would likely affect some people who upgrade. If they did not specify `prune_services: True`, then any configurations added in `firewalld.present` would be added without removing existing services using this version of `firewalld.present`. It also wouldn't prune the other existing firewalld configurations (e.g. rich rules, ports, etc), which it previously would do regardless of the value of `prune_services`. 

However, the old behavior of the function required a complete definition of all preexisting firewall configurations along with the few things that you'd want to change, making it unusable unless you declared all the firewall configurations in a single location. The new behavior when setting these defaults to `True` only requires that you list the configurations you care to add, optionally pruning preexisting configurations only if specified.

We could revert it to its previous default of `False`. To keep the new `prune_` parameters consistent, those should probably be `False` by default as well. However, this would lead to more verbose state functions when you simply want to add services, rich rules, etc without affecting currently present items. 

Do you have any suggestions on how to add the new functionality that supports modular firewall configurations while leaving existing implementations of the function unaffected?

connordelacruz replied:
Hey @cachedout, just wanted to follow up. Any suggestions on the above comments?

gtmanfred replied:
We should not just outright change this, because people are using this as is.

We could change it to None, and if it is None, log a deprecation message that says it defaults to True, but that will be changed to False in 2 releases in the future.

cachedout commented on salt/states/firewalld.py:
```diff
@@ -288,15 +337,21 @@ def service(name,
+            prune_port_fwd=False,
             services=None,
-            prune_services=True,
+            prune_services=False,
```
Same question about upgrades with existing states here.

connordelacruz replied:
See above response

cachedout reviewed (commented):
I have some questions about the upgrade path here. They're in-line. Thanks.

cachedout reviewed (changes requested):
Hi @connordelacruz. Thanks again for this. We do need to hear back on the questions we asked earlier before we can move forward with this, though. Thanks.

connordelacruz committed:
Fixed saltstack/salt#43912

the functions add_port() and add_port_fwd() were enabling masquerade on
the zone if it wasn't enabled already. Neither of these functions
document this behavior and it causes issues when you want to keep
masquerading disabled but want to add ports or port forwards to a
zone. I removed these 2 lines from both functions to fix the issue.

Also removed an unnecessary line from the firewalld state module.
masquerade defaults to False, so the line `masquerade = masquerade or
False` doesn't do anything

connordelacruz commented:
I found an issue (see #43912) with the `firewalld` execution module that affects the state module as well. I fixed the issue in my fork and updated the details of this pull request accordingly.

gtmanfred commented on salt/modules/firewalld.py:
```diff
@@ -629,8 +629,6 @@ def add_port(zone, port, permanent=True):
         salt '*' firewalld.add_port internal 443/tcp
     '''
-    if not get_masquerade(zone):
-        add_masquerade(zone)
```
I don't know that we should just outright change this behavior.  We should probably have an argument passed that disables it.

It is possible that people are using this feature as is.

connordelacruz replied:
As far as I can tell, this behavior is undocumented and seems to be included accidentally, as it doesn't look like it has anything to do with the `add_port` and `add_port_fwd` functions. From what I understand, the module is supposed to be a wrapper for `firewall-cmd`, the CLI for firewalld. The [firewall-cmd documentation](http://www.firewalld.org/documentation/man-pages/firewall-cmd.html) doesn't show that the `--add-port` or `--add-forward-port` flags should affect masquerading, so it seems incongruous with the expected behavior. Additionally, the `firewalld` execution module has a function `add_masquerade` that enables masquerading, so no functionality is removed by making this change.  
We could leave it in and add an argument to skip it, but at that point it feels like we're leaving in a bug and adding an argument to fix the bug.

gtmanfred replied:
The problem is that it is in there, and changing it could change the behavior.

It should be another argument that defaults to True, but can have the default changed in 2 release version, so you can use the salt.utils.warn_until, and mark it for Neon, and we will go ahead and make the change once Fluoride release is out.

connordelacruz replied:
When should `salt.utils.warn_until` be called? (i.e. any time the argument isn't specified?)

gtmanfred replied:
do it basically like you did for the prune services.

set it to True, if None is passed, and then do the warn_until. there.

connordelacruz committed:
Reverted prune_services default behavior and added deprecation warning

- prune_services default value is None
- if prune_services is set to None, _present() sets it to True but logs
a warning message stating that this behavior is deprecated and will be
False in a future release

gtmanfred commented on salt/states/firewalld.py:
```diff
@@ -580,6 +583,11 @@ def _present(name,
+    # if prune_services == None, set to True and log a deprecation warning
+    if prune_services is None:
+        prune_services = True
+        log.warning('prune_services currently defaults to True, but the default will be changed to False in a future release.')
```
You will need to use the `salt.utils.warn_until` function

https://docs.saltstack.com/en/latest/topics/development/deprecations.html

connordelacruz replied:
My bad, I'll use the `salt.utils.warn_until` function instead. would 'Fluorine' be the version to specify in the first argument?

connordelacruz committed:
Using salt.utils.warn_until to log prune_services deprecation warning

connordelacruz committed:
Various changes to handle deprecation with these fixes (see below)

modules/firewalld.py:
- add_port() and add_port_fwd() have a new optional parameter
force_masquerade. When set to True, use old behavior where masquerading
is forcibly enabled in either function. When not specified, a
deprecation warning is logged and force_masquerade is set to True

states/firewalld.py:
- Moved prune_services deprecation warning to present() right before the
call to _present() to make it easier to identify. prune_services is
still set to True if it is unspecified before calling _present()
- Updated ports and port_fwds sections to account for the new
force_masquerade parameter in the execution module.

Additionally, added comments near sections that will be affected by the
deprecation of force_masquerade and prune_services.

connordelacruz committed:
Merge branch 'develop' into develop

connordelacruz committed:
Fixed misleading output for interfaces, sources, and rich rules

Output for these items when not pruning suggested that the old items
were removed. Now output shows added items in addition to old items if
prune_* is False

cachedout commented:
@connordelacruz We're waiting for some responses on a few things here, just so you know. Let us know when you have a moment to look over the latest review comments and we can move forward. Thanks!

connordelacruz committed:
Fixed misleading output for icmp types, ports, port fwds, and services

Output for these items when not pruning suggested that the old items
were removed. Now output shows added items in addition to old items if
prune_* is False

connordelacruz reviewed (commented):
@cachedout I've applied the requested changes in both of the pending reviews. Let me know if there's anything else you'd like me to fix up. Thanks!

connordelacruz commented:
@cachedout @gtmanfred I've made the requested changes for each of the reviews. Let me know when you've had a chance to look them over. Thanks!

nhavens commented:
@connordelacruz Thank you for investing the time in fixing `firewalld.present`. I'm using your code in my _states and _modules directories, and it finally makes `firewalld.present` usable. Prior to your fix, I had to resort to fiddling with the xml zone files with augeas, which was really ugly.

@cachedout, @gtmanfred Do you require anything further from @connordelacruz? I'm eager to have this fix merged in. The current firewalld.present state function has undocumented, arguably unintended side-effects that this pull request resolves. Only with this PR does firewalld.present work. I'd go so far as to say that the deprecation warning in this PR is unnecessary, as the current behavior is broken.

damon-atkins commented:
I looked at this a long time ago, but never got around to fixing it.   With firewalld it allows you to open ports with in the firewalld state. Or enable a xml firewall rule. I ended up creating the xml files first.    However a solution would be to create salt_state_<state_name>.xml files and enable them. 
e.g. /etc/firewalld/services/salt_state_<state-name>.xml
```
<?xml version="1.0" encoding="utf-8"?>
<service>
<short><state-name here></short>
<description>Saltstack communication ports</description>
<port protocol="tcp" port="4505"/>
<port protocol="tcp" port="4506" />
</service>
```
If the state needs to be absent just disable and remove the file.

connordelacruz committed:
Fixed a bug with change output when not pruning

Union operator was being used on a set and a list. Cast the list to a
set, then cast the resulting set from the union back into a list before
adding it to ret['changes']
